parla/task_runtime.py

```diff
@@ -515,20 +536,65 @@ def run(self):
+                        #            When we notify dependees, we don't know places on
+                        #            which the dependees will run since those would decide
+                        #            at the next step, mapping step.
+                        #            Therefore, the current runtime does not consider
+                        #            dependees' placements, but just creates events
+                        #            on the devices where the current task is running.
+                        #            For example, when CPU tasks get GPU events,
+                        #            they will just skip and will not wait for them.
+                        #            This is not technical limitation, just need refactoring
+                        #            and adding more features.
+                        #            But our target applications do not have this pattern
+                        #            and will do it later.
+                        env.wait_dependent_events(self.dependent_events)
+                    # Already waited all dependent events.
+                    # Initialize the dependent event list.
+                    self.dependent_events = []
                     task_state = self._state.func(self, *self._state.args)
+                    env.record_events()
+                    if len(self.events) > 0:
+                        # If any event created by the current task exist,
+                        # notify dependees and make them wait for that event,
+                        # not Parla task completion.
+                        self._notify_dependees_mutex(self.events)
+                    env.sync_events()
                 if task_state is None:
                     task_state = TaskCompleted(None)
             except Exception as e:
                 task_state = TaskException(e)
                 logger.exception("Exception in task")
             finally:
                 logger.info("Finally for task %r", self)
-                # Deallocate all the resources, both from the allocation above and from the "assignment" done by
-                # the scheduler.
+                # Deallocate all the resources, both from the allocation above
+                # and from the "assignment" done by the scheduler.
                 for d in self.req.devices:
-                    ctx.scheduler._available_resources.deallocate_resources(d, self.req.resources)
+                    ctx.scheduler._available_resources. \
+                              deallocate_resources(d, self.req.resources)
+                # Regardless of the previous notification,
+                # (So, before leaving the current run(), the above)
+                # it should notify dependees since
+                # new dependees could be added after the above
+                # notifications, while other devices are running
+                # their kernels asynchronously.
+                self._notify_dependees_mutex()
```

Comment:
I'm concerned about a potential race condition here where a task spawns after this notify but before the task state is set to finished... Is there a way to do this and the next line atomically?

Reply:
Or how about merging the two functions within one mutex statement?
For example, 
``` 
with self.mutex:
     self._notify_dependees()
     self._set_state(task_state)
```

Could it be reasonable?

Reply:
IIUC, the mutex doesn't make it atomic, it just protects the data structures. If dependees checking the task's state must acquire that same mutex to access the state, I think this will work.

Reply:
Yes, generally this is atomic vs transaction. I just suggested this because I have no clue on Python atomic :-) It should share the data strucutre iiuc. So if you think it is fine, we are on board. Let me add this and merge it. Thank you for ur suggestion